cryptoprotocol/src/test/java/org/cryptimeleon/incentivesystem/cryptoprotocol/ProtocolIntegrationTest.java

```diff
@@ -0,0 +1,195 @@
+
+        /**
+         * incentive system setup and user+provider key generation
+         */
+
+        System.out.println("Setting up the incentive system and generating keys.");
+
+        // generate incentive system pp and extracts used Zn for shorter references
+        var incSys = new IncentiveSystem(IncentiveSystem.setup(128, Setup.BilinearGroupChoice.Debug));
+        var usedZn = incSys.getPp().getBg().getZn();
+
+        // generate provider keys
+        var pkp = incSys.generateProviderKeys();
+
+        // generate user key pair for user
+        var ukp = incSys.generateUserKeys();
+
+        /**
+         * user joins system using issue-join protocol
+         */
+
+        System.out.println("A new user joins the system.");
+
+        // user generates join request
+        var joinRequest = incSys.generateJoinRequest(
+                incSys.getPp(),
+                pkp.getPk(),
+                ukp
+        );
+
+        // serialize and deserialize join request to ensure serialization does not break anything
+        var serializedJoinRequest = joinRequest.getRepresentation();
+        FiatShamirProofSystem cwfProofSystem = new FiatShamirProofSystem(new CommitmentWellformednessProtocol(incSys.getPp(), pkp.getPk()));
+        var deserializedJoinRequest = new JoinRequest(serializedJoinRequest, incSys.getPp(), ukp.getPk(), cwfProofSystem);
+
+        // provider handles join request and generates join response
+        var joinResponse = incSys.generateJoinRequestResponse(incSys.getPp(), pkp, ukp.getPk().getUpk(), deserializedJoinRequest);
+
+        // serialize and deserialize join response
+        var serializedJoinResponse = joinResponse.getRepresentation();
+        var deserializedJoinResponse = new JoinResponse(serializedJoinResponse, incSys.getPp());
+
+        // user handles join response
+        var initialToken = incSys.handleJoinRequestResponse(incSys.getPp(), pkp.getPk(), ukp, joinRequest, deserializedJoinResponse);
+
+        /**
+         * transaction 1: user tries to spend points with an empty token
+         */
+
+        System.out.println("Testing spend transaction with empty token.");
+
+        // generate a fresh ID for the spend transaction
+        var tid1 = usedZn.getUniformlyRandomElement();
+
+        var spendAmount1 = BigInteger.ONE;
+
+        // user tries to generate spend request
+        try
+        {
+            var spendRequest1 = incSys.generateSpendRequest(initialToken, pkp.getPk(), spendAmount1, ukp, tid1);
+        }
+        catch(IllegalArgumentException e)
+        {
+            System.out.println("Expected exception when trying to spend " + spendAmount1.toString() +" point using token with " + initialToken.getPoints().toString() + " points: " + e.getMessage());
+        }
+
+        // ensure user token still contains 0 points
+        Assertions.assertEquals(initialToken.getPoints(), usedZn.getZeroElement());
```

Comment:
This statement does not do what you intend since these lines will never fail. generateSpendRequest would not change initialToken